File: advanced-programing-lang/week-2/part-3/NetworkSession.cpp
#include <iostream>
#include <string>

struct NetworkSession {
    std::string session_id;
    NetworkSession(std::string id) : session_id(id) {
        std::cout << "[C++] Connection Opened: " << session_id << "\n";
    }
    ~NetworkSession() {
        std::cout << "[C++] Connection Closed (Freed): " << session_id << "\n";
    }
};

int main() {
    // Manually allocated on the heap
    NetworkSession* s1 = new NetworkSession("Session_A");
    NetworkSession* s2 = new NetworkSession("Session_B");

    delete s1; 
    return 0;
}

File: advanced-programing-lang/week-2/part-3/NetworkSession.rs
struct NetworkSession {
    session_id: String,
}

impl Drop for NetworkSession {
    fn drop(&mut self) {
        println!("[Rust] Connection Closed (Dropped): {}", self.session_id);
    }
}

fn main() {
    println!("Entering Scope...");
    // Allocated via Box (heap pointer)
    let s1 = Box::new(NetworkSession { session_id: String::from("Session_A") });
    
    println!("Exiting Scope...");
    // s1 goes out of scope here. Rust automatically drops it safely.
}